Adapt backend usage depending on wasm file executed

d0iasm opened the pull request:
Adapt backend usage depending on wasm file executed in issue #998.
Close #998 

# Description
Add `auto` backend into a runtime-core and use it as a default backend.
The `auto` backend is equivalent to: 
* singlepass if singlepass is enabled and the wasm file size is larger than 10MiB, or singlepass is enable and the target architecture is aarch64.
* cranelift otherwise.

Hywan commented on src/bin/wasmer.rs:
```diff
@@ -492,6 +492,20 @@ fn execute_wasm(options: &Run) -> Result<(), String> {
         None
     };
 
+    // Update backend when a backend flag is `auto` or a default value.
```
> or a default value

Is it still true? By reading the code, I assume it's when the `--backend` option is set to `auto` only?

d0iasm replied:
I overlooked the Backend::default(). I updated the return value from `Cranelift` to `Auto` in Backend::default(), so this comment should be true.

Hywan commented on src/bin/wasmer.rs:
```diff
@@ -492,6 +492,20 @@ fn execute_wasm(options: &Run) -> Result<(), String> {
     };
 
+    // Update backend when a backend flag is `auto` or a default value.
+    // Use a singlepass if it's enable and the file provided is larger
```
```diff
- a singlepass
+ the Singlepass backend
```

Hywan replied:
```diff
- it's enable
+ it's enabled
```

Hywan commented on src/bin/wasmer.rs:
```diff
@@ -492,6 +492,20 @@ fn execute_wasm(options: &Run) -> Result<(), String> {
 
+    // Update backend when a backend flag is `auto` or a default value.
+    // Use a singlepass if it's enable and the file provided is larger
+    // than 10MiB (10485760 bytes), or it's enable and the target architecture
```
```diff
- it's enable
+ it's enabled
```

Hywan commented on src/bin/wasmer.rs:
```diff
@@ -492,6 +492,20 @@ fn execute_wasm(options: &Run) -> Result<(), String> {
+    // Update backend when a backend flag is `auto` or a default value.
+    // Use a singlepass if it's enable and the file provided is larger
+    // than 10MiB (10485760 bytes), or it's enable and the target architecture
+    // is aarch64. Otherwise, use a cranelift as a backend.
```
```diff
- use a cranelift as a backend
+ use the Cranelift backend
```

Hywan commented on src/bin/wasmer.rs:
```diff
@@ -492,6 +492,20 @@ fn execute_wasm(options: &Run) -> Result<(), String> {
+    // is aarch64. Otherwise, use a cranelift as a backend.
+    if options.backend == Backend::Auto {
+        if Backend::variants().contains(&Backend::Singlepass.to_string())
+            && (wasm_binary.len() > 10485760 || cfg!(target_arch = "aarch64"))
```
Based on my comprehension of #998, when the `--backend` option is set to `auto`, we want to use Cranelift by default, or Singlepass if the module is larger than 10MiB; which translates as:

```rust
if Backend::variants().containts(&Backend::Singlepass.to_string()) // Singlepass is available
    && wasm_binary.len() > 10485760 // Large module, be a good candidate for Singlepass
{
    options.backend = Backend::Singlepass;
} else {
    options.backend = Backend::Cranelift;
}
```

d0iasm replied:
The issue #998 only mentions the file size, but I heard from @syrusakbary on Slack that the  Singlepass backend is used when the target arch is AArch64. Is it better to add about it to the issue #998, or to remove the condition relating to AArch64 for now?

Hywan commented:
@d0iasm Don't forget to update the `CHANGELOG.md` file too :-).

d0iasm commented:
@MarkMcCaskey 
Thank you for your suggestions! I chose option 1 because option 2 would affect many places since `options` is used in other functions such as `execute_wasi`. I added a new function `update_backend` called from `run`. I also reverted `options: &mut Run` to `options: &Run` in `execute_wasm` argument.

Hywan commented:
bors r+

Hywan commented:
bors r+

d0iasm commented:
bors r+

d0iasm commented:
The test `cache::tests::test_file_system_cache_run` fails because the default compiler is auto but there is no "real" compiler of auto. So, I reverted `Backend::default()` to return `Backend::Cranelift`.

d0iasm commented:
bors try

MarkMcCaskey commented:
bors r+
